pkg/network/common/dns.go

```diff
@@ -101,34 +125,69 @@ func (d *DNS) Delete(dns string) {
 	delete(d.dnsMap, dns)
 }
 
-func (d *DNS) Update(dnsName string) (bool, error) {
+func (d *DNS) LockForUpdate(dns string) error {
 	d.lock.Lock()
 	defer d.lock.Unlock()
-
-	return d.updateOne(dnsName)
-}
-
-func (d *DNS) updateOne(dns string) (bool, error) {
 	res, ok := d.dnsMap[dns]
 	if !ok {
 		// Should not happen, all operations on dnsMap are synchronized by d.lock
-		return false, fmt.Errorf("DNS value not found in dnsMap for domain: %q", dns)
+		return fmt.Errorf("DNS value not found in dnsMap for domain: %q", dns)
 	}
 
+	res.updating = true
+	d.dnsMap[dns] = res
+
+	return nil
+}
+
+func (d *DNS) Update(dns string) (bool, error) {
+	// This is a blocking operation, therefore must be done before acquring
+	// the lock
 	ips, ttl, err := d.getIPsAndMinTTL(dns)
+
+	defer func(start time.Time) {
+		warnSlowOperation("Update egress DNS record after resolution", start)
+	}(time.Now())
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
 	if err != nil {
-		res.nextQueryTime = time.Now().Add(defaultTTL * time.Second)
-		d.dnsMap[dns] = res
+		klog.Errorf("Unable to resolv %q: %v", dns, err)
+		err = d.updateNextQueryTime(dns)
 		return false, err
 	}
 
+	return d.updateDNSValue(dns, ips, ttl)
+}
+
+func (d *DNS) updateNextQueryTime(dns string) error {
+	res, ok := d.dnsMap[dns]
+	if !ok {
+		// Should not happen, all operations on dnsMap are synchronized by d.lock
+		return fmt.Errorf("DNS value not found in dnsMap for domain: %q", dns)
+	}
+	res.nextQueryTime = time.Now().Add(res.ttl)
+	res.updating = false
+	d.dnsMap[dns] = res
+	return nil
+}
+
+func (d *DNS) updateDNSValue(dns string, ips []net.IP, ttl time.Duration) (bool, error) {
+	res, ok := d.dnsMap[dns]
+	if !ok {
+		// Should not happen, all operations on dnsMap are synchronized by d.lock
+		return false, fmt.Errorf("DNS value not found in dnsMap for domain: %q", dns)
+	}
+
 	changed := false
 	if !ipsEqual(res.ips, ips) {
 		changed = true
 	}
 	res.ips = ips
 	res.ttl = ttl
 	res.nextQueryTime = time.Now().Add(res.ttl)
+	res.updating = false
```

Comment:
why do it twice if we are also doing it as a part of `updateNextQueryTime`?

Reply:
We shouldn't do it twice for the same update. If `getIPsAndMinTTL` returns an error we call `updateNextQueryTime`, else we call `updateDNSValue`.